pkg/imgpkg/image/keychain.go

```diff
@@ -0,0 +1,208 @@
+	case len(k.Token) > 0:
+		return &regauthn.Bearer{Token: k.Token}, nil
+	case k.Anon:
+		return regauthn.Anonymous, nil
+	default:
+		return retryDefaultKeychain(func() (regauthn.Authenticator, error) {
+			return regauthn.DefaultKeychain.Resolve(res)
+		})
+	}
+}
+
+func retryDefaultKeychain(doFunc func() (regauthn.Authenticator, error)) (regauthn.Authenticator, error) {
+	var auth regauthn.Authenticator
+	var lastErr error
+
+	for i := 0; i < 5; i++ {
+		auth, lastErr = doFunc()
+		if lastErr == nil {
+			return auth, nil
+		}
+
+		if strings.Contains(lastErr.Error(), errCredentialsNotFoundMessage) || strings.Contains(lastErr.Error(), errCredentialsMissingUsernameMessage) || strings.Contains(lastErr.Error(), errCredentialsMissingServerURLMessage) {
+			return auth, lastErr
+		}
+
+		time.Sleep(2 * time.Second)
+	}
+	return auth, fmt.Errorf("Retried 5 times: %s", lastErr)
+}
+
+type envKeychain struct {
+	globalPrefix string
+
+	infos       []envKeychainInfo
+	collectErr  error
+	collected   bool
+	collectLock sync.Mutex
+}
+
+var _ regauthn.Keychain = &envKeychain{}
+
+func (k *envKeychain) Resolve(target regauthn.Resource) (regauthn.Authenticator, error) {
+	infos, err := k.collect()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, info := range infos {
+		if info.Hostname == target.RegistryStr() {
+			return regauthn.FromConfig(regauthn.AuthConfig{
+				Username:      info.Username,
+				Password:      info.Password,
+				IdentityToken: info.IdentityToken,
+				RegistryToken: info.RegistryToken,
+			}), nil
+		}
+	}
+
+	return regauthn.Anonymous, nil
+}
+
+func (k *envKeychain) collect() ([]envKeychainInfo, error) {
+	k.collectLock.Lock()
+	defer k.collectLock.Unlock()
+
+	if k.collected || k.collectErr != nil {
+		return append([]envKeychainInfo{}, k.infos...), k.collectErr
+	}
+
+	const (
+		sep = "_"
+	)
+
+	funcsMap := map[string]func(*envKeychainInfo, string) error{
+		"HOSTNAME": func(info *envKeychainInfo, val string) error {
+			registry, err := regname.NewRegistry(val, regname.StrictValidation)
+			if err != nil {
+				return fmt.Errorf("Parsing registry hostname: %s (e.g. gcr.io, index.docker.io)", err)
+			}
+			info.Hostname = registry.RegistryStr()
+			return nil
+		},
+		"USERNAME": func(info *envKeychainInfo, val string) error {
+			info.Username = val
+			return nil
+		},
+		"PASSWORD": func(info *envKeychainInfo, val string) error {
+			info.Password = val
+			return nil
+		},
+		"IDENTITY_TOKEN": func(info *envKeychainInfo, val string) error {
+			info.IdentityToken = val
+			return nil
+		},
+		"REGISTRY_TOKEN": func(info *envKeychainInfo, val string) error {
+			info.RegistryToken = val
+			return nil
+		},
+	}
+
+	defaultInfo := envKeychainInfo{}
+	infos := map[string]envKeychainInfo{}
+
+	for _, env := range os.Environ() {
+		pieces := strings.SplitN(env, "=", 2)
+		if len(pieces) != 2 {
+			continue
+		}
+
+		var matched bool
+
+		for key, updateFunc := range funcsMap {
```

Comment:
Since `funcsMap` is a Map I think that there should be a more efficient way to do this